File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if (s1.size() > s2.size()) {
            return false;
        }
        
        vector<int> s1Map(26), s2Map(26);
        
        for (int i = 0; i < s1.size(); i++) {
            s1Map[s1[i] - 'a']++;
            s2Map[s2[i] - 'a']++;
        }
        
        for (int i = 0; i < s2.size() - s1.size(); i++) {
            if (s1Map == s2Map) {
                return true;
            }
            
            s2Map[s2[i] - 'a']--;
            s2Map[s2[i + s1.size()] - 'a']++;
        }
        
        return s1Map == s2Map;
    }
};

File: 0567-permutation-in-string/0567-permutation-in-string.cs
public class Solution {
    private static bool check(int[] freq1, int[] freq2) {
        for (int i=0; i<26; ++i)
            if (freq1[i] < freq2[i]) return false;
        return true;
    }
    public bool CheckInclusion(string s1, string s2) {
        int[] freq1 = new int[26];
        foreach (char x in s1) freq1[x - 'a']++;
        int[] freq2 = new int[26];
        int left = 0, right = 0;
        while (right < s2.Length) {
            freq2[s2[right] - 'a']++;
            if (right >= s1.Length - 1) {
                if (check(freq1, freq2)) return true;
                freq2[s2[left++] - 'a']--;
            }
            right++;
        }
        return false;
    }
}